apps/src/netsim/NetSimRouterNode.js

```diff
@@ -948,10 +1110,149 @@ NetSimRouterNode.prototype.routeMessage_ = function (message, onComplete) {
 };
 
 /**
+ * @param {messageRow} messageRow
+ * @returns {boolean} TRUE if the given row should be operated on by the local
+ *          simulation, FALSE if another user's simulation should handle it.
+ * @private
+ */
+NetSimRouterNode.prototype.localSimulationOwnsMessageRow_ = function (messageRow) {
+  // Local simulation can't handle anything if it's not initialized.
+  if (!this.simulateForSender_) {
+    return false;
+  }
+
+  var packet, fromAddress, toAddress, simulateForAddress;
+  var routerNodeID = this.entityID;
+
+  // One extra responsibility in auto-DNS mode: Should handle messages from the
+  // auto-DNS to the router IF they are addressed to the local client.
+  if (this.dnsMode === DnsMode.AUTOMATIC) {
+
+    // Gather address info to confirm this packet's destination
+    packet = new Packet(this.packetSpec_, messageRow.payload);
+    fromAddress = packet.getHeaderAsInt(Packet.HeaderType.FROM_ADDRESS);
+    toAddress = packet.getHeaderAsInt(Packet.HeaderType.TO_ADDRESS);
+    simulateForAddress = this.getAddressForNodeID_(this.simulateForSender_);
+
+    var isOnRouterLoopback = messageRow.fromNodeID === routerNodeID &&
+        messageRow.toNodeID === routerNodeID;
+
+    var isBetweenLocalAndDns = (
+        fromAddress === simulateForAddress &&
+        toAddress === AUTO_DNS_RESERVED_ADDRESS
+        ) || (
+        fromAddress === AUTO_DNS_RESERVED_ADDRESS &&
+        toAddress === simulateForAddress
+        );
+
+    if (isOnRouterLoopback && isBetweenLocalAndDns) {
+      return true;
+    }
+  }
+
+  // Local simulation of router normally handles only the messages from the
+  // local client to the router.
+  return messageRow.fromNodeID === this.simulateForSender_ &&
+      messageRow.toNodeID === routerNodeID;
+};
+
+/**
+ * @param {NetSimMessage} message
+ * @returns {number} time required to process this message, in milliseconds.
+ * @private
+ */
+NetSimRouterNode.prototype.calculateProcessingDurationForMessage_ = function (message) {
+  if (this.bandwidth === Infinity) {
+    return 0;
+  }
+  return message.payload.length * 1000 / this.bandwidth;
+};
+
+/**
+ * Update queue of all auto-dns messages, which can be used for stats or limiting.
+ * @param {messageRow[]} rows
+ * @private
+ */
+NetSimRouterNode.prototype.updateAutoDnsQueue_ = function (rows) {
+  var newQueue = rows.filter(this.isMessageToAutoDns_.bind(this));
+  if (!_.isEqual(this.autoDnsQueue_, newQueue)) {
```

Comment:
Could invert if there's going to be a lot of content in this block in the future